### ISSUE TITLE
Range: bytes header for warp::fs::dir hosted files doesn't receive last chunk

### ISSUE THREAD

**Issue by jaredwolff** [seanmonstar/warp#815]
**Version**
warp v0.3.0

**Platform**
FreeBSD pyrinas-dev.jaredwolff.com 12.2-RELEASE-p1 FreeBSD 12.2-RELEASE-p1 GENERIC  amd64

**Description**

Not sure if this should go here or to the underlying `hyper` implementation but I'll give it a shot here first.

I'm currently hosting some static binaries that can be downloaded by embedded devices for OTA updates. One of the important pieces of functionality is the need for partial downloads (say 2048 byte chunks at a time). While this, mostly, works, I've run into an issue related to retrieving the last chunk. 

For instance I know that the file located at  `https://pyrinas-dev-ota.jaredwolff.com/images/352656102712240.bin` is 200595 bytes in size. (Btw, it's a test binary on publicly available code)

When I attempt to pull the last section using curl I get a 416 error: (I originally got this error in embedded which led me to investigate more.)

```
curl/7.73.0/bin/curl https://pyrinas-dev-ota.jaredwolff.com/images/352656102712240.bin -H "Content-Type: application/octet-stream" -H "Range: bytes=198656-200595" --output bob.bin
```

But, when I reduce the size of the full file, it retrieves the last 1939 bytes successfully. 

```
/usr/local/Cellar/curl/7.73.0/bin/curl https://pyrinas-dev-ota.jaredwolff.com/images/352656102712240.bin -H "Content-Type: application/octet-stream" -H "Range: bytes=198656-200594" --output bob.bin
```

I am running this through a reverse proxy (Caddy). I did double check that the Caddy implementation is not funky by doing the same thing with a statically hosted file elsewhere. (Same Caddy instance, different domain) No problems though with the .png I was hosting.

```
curl  https://www.jaredwolff.com/testing-hardware-using-rust/images/Screen_Shot_2021-01-06_at_10.30.55_AM.png -H "Range: bytes=125801-125808" --output test.bin
```

All the above files are available for anyone to test this. Seems like the implementation is using a Range indexed from 0 rather than from 1 which may explain why reducing to 200594 worked.

By the way here's what the code to host the files looks like:

```rust
    // Only one folder that we're interested in..
    let images = warp::path("images").and(warp::fs::dir(&IMAGE_FOLDER_PATH));

    // Run the `warp` server
    warp::serve(images)
        .run(SocketAddrV4::new(
            Ipv4Addr::new(127, 0, 0, 1),
            settings.http_port,
        ))
        .await;
```

Please let me know if you need any more details.

**Comment by jaredwolff** [comment]
I also tried this without any HTTPS in between directly with the `warp` based server. Same results. 

```
#
curl 10.10.2.41:3030/images/352656102712240.bin -H "Content-Type: application/octet-stream" -H "Range: bytes=198656-200595" --output bob.bin
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
# 
curl 10.10.2.41:3030/images/352656102712240.bin -H "Content-Type: application/octet-stream" -H "Range: bytes=198656-200594" --output bob.bin
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:-- 100  1939  100  1939    0     0  43088      0 --:--:-- --:--:-- --:--:-- 44068
```

**Comment by jaredwolff** [comment]
Actually this *should* be expected behavior in a server. Looks like the embedded client I'm using is inadvertently requesting 1 byte too many since the range *does* start at 0.

**Comment by jaredwolff** [comment]
Actually trying this on a few other servers, they seem to handle the case where the amount on the right is the final size in bytes. (Rather than _file_size -  1_  I've tried this on an AWS backend, Github with a random picture file, and, as mentioned before, serving a static file on my site via Caddy.

**Comment by jaredwolff** [comment]
I can see in the `headers` crate by defaults bound are included by default: https://github.com/hyperium/headers/blob/c20737231dc7a1cdcd787dc3249aaa0992244881/src/common/range.rs#L83

Whereas in the bytes_range it looks like included gets a + 1 by default even though it may be fed the full file size:
https://github.com/seanmonstar/warp/blob/68ac3a595a1969347df6ef66ca327d315b12ab7d/src/filters/fs.rs#L358

Maybe change the `Included` option where it would look something like:

```
if s == max_len {
   s
} else {
   s + 1
}
```

This way when the full file length is fed here, it won't have an error as end will be `<= max_len`

**Comment by jxs** [comment]
fixed with #818
